### ISSUE TITLE
Ambiguity on the generation of corruptions

### ISSUE THREAD

**Issue by AlexisPister** [Accenture/AmpliGraph#84]
Hello,

When I fit a model, specifiyng a list of entities for the corruption ("corruption_entities" parameter in the "early_stopping_params" dictionnary), I always get the following message which is contradictory :

"ampligraph.latent_features.models - INFO - Using all entities for generation of corruptions"

I get the same thing when I just call model.fit without the early stopping heuristic. What is the expected behaviour of the model when there is no early stopping parameters ? Are corrupted entities still generated ?

**Comment by chanlevan** [comment]
Hi @AlexisPister 

Thanks for letting us know this. It's a bit confusion indeed but this is merely a logging matter. 

There are two places used to generate the corruption entities, 1 can be configured when you initialize a model through `embedding_model_params={'negative_corruption_entities': 'all'/'batch'/list}`, and 1 in early stopping through `early_stopping_params = {'corruption_entities': 'all'/'batch'/list}`

What you saw was the default behaviour when you are not passing the `embedding_model_params={'negative_corruption_entities': 'all'/'batch'/list}` to when you initialise the model and there was no logging information of early stopping regarding to that in the first release. 

From commit `8cffba8`, we have added more logging information when you use early stopping. By default, there will no more logging shown. If you want to see it, you can go to ampligraph/logger.conf and change `handler_consoleHandler` logging level to `DEBUG`.

For example, with below code snippet:

`import numpy as np
from ampligraph.datasets import load_wn18
from ampligraph.latent_features import ComplEx

def main():
    X = load_wn18()
    model = ComplEx(batches_count=256, seed=0, epochs=20, k=150, eta=10,
                    optimizer='adam', optimizer_params={'lr':1e-3},
                    loss='pairwise', loss_params={'margin':0.5},
                    regularizer='LP', regularizer_params={'p':2, 'lambda':1e-5}, 
                    verbose=True)

    filter = np.concatenate((X['train'], X['valid'], X['test']))
    
    model.fit(X['train'], 
              early_stopping = True,
              early_stopping_params = \
                      {
                          'x_valid': X['valid'],       
                          'criteria':'hits10',         
                          'burn_in': 100,              
                          'check_interval':20,         
                          'stop_interval':5,           
                          'x_filter': filter,          
                          'corruption_entities':[1,2,3,4,5], 
                          'corrupt_side':'s+o'         
                      }
              )
if __name__ == "__main__":
    main()`

You will see the output like this:
`...
DEBUG - Using all entities for generation of corruptions during training
DEBUG - Requires the same size of postive and negative
DEBUG - Generating corruptions for fit.
DEBUG - Created corruption masks.
DEBUG - Created corrupted subjects.
DEBUG - Retained relationships.
DEBUG - Created corrupted objects.
DEBUG - Returning corruptions for fit.
DEBUG - Creating dependencies before loss computations.
DEBUG - Dependencies found:
	Required same size positive and negative.
	Eta is not 1.
DEBUG - Converting statements to integer ids.
DEBUG - Using the supplied entities for generation of corruptions for early stopping
DEBUG - Converting statements to integer ids.
DEBUG - Reading from prime_number_list.txt.
...`

**Comment by AlexisPister** [comment]
Oh ok, it's all clear now, I did not understand there were 2 places to generate corrupted entities ! Thank you.

Moreover, does the doc explain how the early stop heuristic work ? I did not find any page.

Thanks

**Comment by lukostaz** [comment]
When `early_stopping=True`, the training loop adopts the following early stopping heuristics:
- the model will first train for `burn_in` epochs.
- then, every `check_interval` epochs it will compute the metric specified in `criteria`. If such metric decreases for `stop_interval` checks, we stop training early.
- Note the metric is computed on `x_valid`. This is usually a validation set that you held out. Also, because `criteria` is a ranking metric, the procedure also requires negatives. We decided to give users quite a lot of flexibility, so you can specify what are the entities to use to generate corruptions, and which side(s) of a triple you want to corrupt. You can also compute filtered metrics, by passing an array of positives that will be used to filter your corruptions (use `x_filter` for that).

Keep in mind the early stopping criteria may introduce a certain overhead (caused by the metric computation). The goal is to strike a good trade-off between such overhead and saving training epochs. 

A common approach uses MRR unfiltered:

```
early_stopping_params={x_valid=X['valid'], 'criteria': 'mrr'}
```

Note the size of validation set also contributes to such overhead. In most cases a smaller validation set would be enough.

**Comment by lukostaz** [comment]
Additional documentation added in `5a418088c4577f9a01c4e60941b9aa2f89ccf46f`. Closing.
